examples/machine_translation/train.py

```diff
@@ -0,0 +1,119 @@
+        loss = criterion(output, trg)
+
+        loss.backward()
+        torch.nn.utils.clip_grad_norm_(model.parameters(), clip)
+        optimizer.step()
+
+        epoch_loss += loss.item()
+    return epoch_loss / len(iterator)
+
+
+def evaluate(model: nn.Module, iterator: DataLoader, criterion: nn.Module):
+    model.eval()
+
+    epoch_loss = 0
+    with torch.no_grad():
+        for _, batch in enumerate(iterator):
+            # Need to convert to [seq_len x batch x ...] size
+            src = batch[0].transpose(0, 1)
+            trg_char = batch[1][0].transpose(0, 1)
+            trg_word = batch[1][1].transpose(0, 1)
+
+            output = model(src, trg_char, 0)  # turn off teacher forcing
+            output = output[1:].reshape(-1, output.shape[-1])
+            trg = trg_word[1:].reshape(-1)
+
+            loss = criterion(output, trg)
+
+            epoch_loss += loss.item()
+    return epoch_loss / len(iterator)
+
+
+if __name__ == "__main__":
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    batch_size = 32
+    train_dataset, val_dataset, test_dataset = get_dataset()
+    train_iterator = DataLoader(train_dataset, batch_size=batch_size, collate_fn=collate_fn)
+    valid_iterator = DataLoader(val_dataset, batch_size=batch_size, collate_fn=collate_fn)
+    test_iterator = DataLoader(test_dataset, batch_size=batch_size, collate_fn=collate_fn)
+    input_dim = len(train_dataset.vocab[0])
+    char_output_dim = len(train_dataset.vocab[1])
+    word_output_dim = len(train_dataset.vocab[2])
+
+    # enc_emb_dim = 256
+    # dec_emb_dim = 256
+    # enc_hid_dim = 512
+    # dec_hid_dim = 512
+    # attn_dim = 64
+    # enc_dropout = 0.5
+    # dec_dropout = 0.5
+
+    enc_emb_dim = 300
+    dec_emb_dim = 300
+    enc_hid_dim = 64
+    dec_hid_dim = 64
+    attn_dim = 8
+    enc_dropout = 0.5
+    dec_dropout = 0.5
+
+    enc_emb = WordCharCNNEmbedding(input_dim)
+    enc = Encoder(input_dim, enc_emb_dim, enc_hid_dim, dec_hid_dim, enc_dropout, enc_emb)
+    attn = Attention(enc_hid_dim, dec_hid_dim, attn_dim)
+    dec_emb = WordCharCNNEmbedding(char_output_dim)
+    dec = Decoder(word_output_dim, dec_emb_dim, enc_hid_dim, dec_hid_dim, dec_dropout, attn, dec_emb)
+    model = Seq2Seq(enc, dec, device).to(device)
+
+    criterion = nn.CrossEntropyLoss(ignore_index=train_dataset.vocab[2].stoi["<pad>"])
+    optimizer = optim.Adam(model.parameters())
+    print(f"The model has {count_parameters(model):,} trainable parameters")
+    n_epochs = 10
+    clip = 1
+
+    best_valid_loss = float("inf")
+
+    for epoch in range(n_epochs):
+
+        start_time = time.time()
+
+        train_loss = train(model, train_iterator, optimizer, criterion, clip)
+        valid_loss = evaluate(model, valid_iterator, criterion)
+
+        end_time = time.time()
+
+        epoch_mins, epoch_secs = epoch_time(start_time, end_time)
+
+        print(f"Epoch: {epoch+1:02} | Time: {epoch_mins}m {epoch_secs}s")
```

Comment:
Should we use `bleu_score` for a metric?
https://github.com/pytorch/text/blob/bcb9104680eb9dc978a6bbcc2b9ca46cf2bdbed9/torchtext/data/metrics.py#L35

Reply:
done